sheets/utils.py

```diff
@@ -40,407 +40,194 @@ def generate_google_client_config():
+        self.num_columns = len(self.column_headers)
 
-        Returns:
-            (str or None, datetime.datetime or None): The new status paired with the date that
-                the status was logged in Mailgun (or (None, None) if the given assignment does not
-                have a new status.
-        """
-        message_data = self._assignment_map[bulk_assignment_id].get(
-            (code, recipient_email)
-        )
-        return (
-            (message_data["new_status"], message_data["new_status_date"])
-            if message_data
-            else (None, None)
+        self.LAST_COL_LETTER = get_column_letter(self.num_columns - 1)
+        self.STATUS_COL = (
+            i for i, header in enumerate(self.column_headers) if header == "Status"
         )
 
-    def has_new_statuses(self, bulk_assignment_id):
-        """
-        Returns True if the given bulk assignment has any individual assignments with a new status
 
-        Args:
-            bulk_assignment_id (int):
+request_sheet_metadata = CouponRequestSheetMetadata()
+refund_sheet_metadata = RefundRequestSheetMetadata()
+assign_sheet_metadata = CouponAssignSheetMetadata()
 
-        Returns:
-            bool: True if the given bulk assignment has any individual assignments with a new status
-        """
-        return any(
-            message_data["new_status"] is not None
-            for _, message_data in self._assignment_map[bulk_assignment_id].items()
-        )
 
-    def get_status_date_rows(self, bulk_assignment_id):
-        """
-        Returns a row index, status, and status date for each coupon assignment with a new status in the
-        given bulk assignment
+class ResultType(Enum):
+    """Enum of possible row results"""
 
-        Args:
-            bulk_assignment_id (int):
+    IGNORED = "ignored"
+    FAILED = "failed"
+    OUT_OF_SYNC = "out_of_sync"
+    PROCESSED = "processed"
 
-        Returns:
-            iterable of (int, str, datetime.datetime): An iterable of row indices (indicating the Sheet row) paired
-                with new status and the date of that status change
-        """
-        return (
-            (
-                message_data["row_index"],
-                message_data["new_status"],
-                message_data["new_status_date"],
-            )
-            for message_data in self._assignment_map[bulk_assignment_id].values()
-            if message_data["new_status"]
-        )
+    def __eq__(self, other):
+        return self.value == other.value  # pylint: disable=comparison-with-callable
 
-    def get_sheet_id(self, bulk_assignment_id):
-        """
-        Returns the assignment spreadsheet id associated with the given bulk assignment
+    def __lt__(self, other):
+        return self.value < other.value  # pylint: disable=comparison-with-callable
 
-        Args:
-            bulk_assignment_id (int):
+    def __hash__(self):
+        return hash(self.value)
```

Comment:
I'm not sure we need this. It looks like there is already a hash value for the given enum.

Reply:
Regarding this and the other comment on `__lt__` above, I ran into a problem with sorting based on these enum values. Here's a canned example that raises an exception:

```python
import itertools
from collections import namedtuple
import operator as op
from enum import Enum

class ResultType(Enum):
    """Enum of possible row results"""
    IGNORED = "ignored"
    FAILED = "failed"
    PROCESSED = "processed"
    
RowResult = namedtuple("RowResult", ["row_index", "result_type"])

def group_into_dict(items, key_fn):
    sorted_items = sorted(items, key=key_fn)
    return {
        key: list(values_iter)
        for key, values_iter in itertools.groupby(sorted_items, key=key_fn)
    }

row_results = [
    RowResult(row_index=1, result_type=ResultType.PROCESSED),
    RowResult(row_index=2, result_type=ResultType.IGNORED),
    RowResult(row_index=3, result_type=ResultType.FAILED),
    RowResult(row_index=4, result_type=ResultType.IGNORED),
    RowResult(row_index=5, result_type=ResultType.PROCESSED),
]

group_into_dict(row_results, key_fn=op.attrgetter("result_type"))
```

The exception: `TypeError: '<' not supported between instances of 'ResultType' and 'ResultType'`. For that I added the `__lt__`. I was also running into exceptions with using those enums as dict keys, but now I just tried the same example without the `__hash__` function and it worked, so I guess I'll get rid of that 🤷‍♂